### PR TITLE
Added Password Complexity to register and password reset

### PR DESCRIPTION
**Pull request by trite8q1**
|               Status                |                Type                 | Env Vars Change |
| :---------------------------------: | :---------------------------------: | :-------------: |
| :white_check_mark: Ready | Feature |     No      |

## Description

- Implements validatePassword function in AuthServiceProvider for password complexity
- Enhancement of register and resetPassword functions in AuthController

## Motivation and Context

It implements password complexity.

## Checklist

- [x] I have read the **CONTRIBUTING** document.
- [x] I have considered the accessibility of my changes (i.e. did I add proper content descriptions to images, or run my changes with talkback enabled?)
- [x] I have documented my code if needed

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -47,10 +47,10 @@
     "mariadb": "^2.5.3",
     "morgan": "^1.10.0",
     "mysql2": "^2.2.5",
-    "pg": "^8.5.1",
     "pg-hstore": "^2.3.2",
     "sequelize": "^6.6.2",
     "short-unique-id": "^4.4.4",
+    "sqlite3": "^5.0.2",
```
Comment (noctera) [inline package.json]: Don't change the package dependencies, please

Comment (luwol03) [reply package.json]: Please also remove the package-lock.json from the changes.

File: app/Controllers/AuthController.js [hunk 2]
```diff
@@ -22,7 +23,10 @@ const register = catchAsync(async (req, res) => {
     await useInviteCode(req.query.inviteCode);
   }
 
-  res.send({ token, user });
+  if (validatePassword(req.body.password)) {
+    res.send({ token, user });
+  }
+  res.status(400).end();
```
Comment (luwol03) [inline app/Controllers/AuthController.js]: Can you do that the other way. First check if password is not valid and then `return res.status(400).end()` and then after that if do the res.send thing.

File: app/Controllers/AuthController.js [hunk 3]
```diff
@@ -58,6 +62,9 @@ const resetPassword = catchAsync(async (req, res) => {
 
   await changePassword(userId, oldPassword, newPassword);
 
+  if (validatePassword(req.body.newPassword)) {
+    res.status(200).end();
+  }
   res.status(204).end();
```
Comment (luwol03) [inline app/Controllers/AuthController.js]: Same here

File: app/Services/AuthServiceProvider.js [hunk 4]
```diff
@@ -82,6 +83,25 @@ function validateLogin(req, res) {
...
+  const hasLowerCase = /[a-z]/.test(password);
+  const hasNumbers = /\d/.test(password);
+  const hasNonAlphas = /\W/.test(password);
+
+  // eslint-disable-next-line no-nested-ternary
+  const passwordComplexity = length
+    ? length + hasUpperCase + hasLowerCase + hasNumbers + hasNonAlphas
+    : passwordLength !== 0
+    ? 1
+    : 0;
```
Comment (luwol03) [inline app/Services/AuthServiceProvider.js]: It wasn't the best way from me of doing this. Can you please try to write that with normal if statements? I'll also change that in my pr as well.

**Review by luwol03 — CHANGES_REQUESTED** [review]

**Comment by luwol03** [comment]
Sorry, I can't see discord messages right now due to discord outage. Do you also have such issues?

File: app/Services/AuthServiceProvider.js [hunk 5]
```diff
@@ -82,6 +83,27 @@ function validateLogin(req, res) {
...
+function validatePassword(password) {
+  const passwordLength = bytesLength(password);
+
+  const length = passwordLength >= 8 && passwordLength <= 72;
+  const hasUpperCase = /[A-Z]/.test(password);
+  const hasLowerCase = /[a-z]/.test(password);
+  const hasNumbers = /\d/.test(password);
+  const hasNonAlphas = /\W/.test(password);
+
+  let passwordComplexity;
```
Comment (luwol03) [inline app/Services/AuthServiceProvider.js]: you can directly set the complexity to 0 and remove the else statement below

File: app/utils/index.js [hunk 6]
```diff
@@ -286,6 +286,8 @@ const askToConfirm = (question) => {
   });
 };
 
+const bytesLength = (string) => new TextEncoder().encode(string).length;
```
Comment (luwol03) [inline app/utils/index.js:295]: please also add the comments above of that function. [See](https://github.com/vocascan/vocascan-frontend/pull/97/files#diff-ebd8554ec43b07366bd40d16b09d1d063d9e8d5921fa30de4eb80bffdf82668dR60-R66)

File: app/Controllers/AuthController.js [hunk 7]
```diff
@@ -56,8 +59,10 @@ const resetPassword = catchAsync(async (req, res) => {
   const userId = req.user.id;
   const { oldPassword, newPassword } = req.body;
 
+  if (!validatePassword(req.body.newPassword)) {
+    res.status(400).end();
+  }
   await changePassword(userId, oldPassword, newPassword);
-
```
Comment (luwol03) [inline app/Controllers/AuthController.js:60]: Add that newline back

File: app/Services/AuthServiceProvider.js [hunk 8]
```diff
@@ -82,6 +83,27 @@ function validateLogin(req, res) {
...
+
+  if (length) {
+    passwordComplexity = length + hasUpperCase + hasLowerCase + hasNumbers + hasNonAlphas;
+  } else if (passwordLength !== 0) {
+    passwordComplexity = 1;
+  } else {
+    passwordComplexity = 0;
+  }
+  return passwordLength >= 8 && passwordLength <= 72 && passwordComplexity >= 4;
```
Comment (luwol03) [inline app/Services/AuthServiceProvider.js]: Please also add a newline between those two lines.

**Review by luwol03 — CHANGES_REQUESTED** [review]

**Review by luwol03 — APPROVED** [review]

**Pull request merged**